ttkwidgets/shell.py

```diff
@@ -0,0 +1,205 @@
+import tkinter as tk
+import tkinter.font as tkfont
+from collections import defaultdict
+
+
+class Shell(tk.Canvas):
+    def __init__(self, master, textvariable=None, prefix='', force_focus=True,
+                 font=None, **kwargs):
+        """
+        :param master: parent widget
+        :type master: tkinter.Widget
+        :param textvariable: A tkinter variable that holds the current text buffer
+        :type textvariable: tkinter.StringVar or None
+        :param prefix: A prefix to show on every input line
+        :type prefix: str
+        :param force_focus: whether or not the shell should take the focus.
+        :type force_focus: bool
+        :param font: Font to use for the terminal
+        :type font: tkinter.font.Font, tuple, or None
+        """
+        for key, value in {
+            'background': 'black',
+            'takefocus': True,
+        }.items():
+            if key not in kwargs:
+                kwargs[key] = value
+        super().__init__(master, **kwargs)
+        if kwargs['takefocus'] and force_focus:
+            self.focus_force()
+
+        self.bind('<Key>', self.on_key_press)
+        self.bind('<Configure>', self.on_configure)
+        self.textvariable = textvariable if isinstance(textvariable, tk.StringVar) else tk.StringVar()
+        self.prefix = prefix
+        self.font = font or ('Terminal', 10)
+        self.line_pos = (5, 5)
+        self.texts = []
+        self.last_text = None
+        self.cursor = None
+        self.blink = True
+        self.buffer = prefix
+        self.text_update()
+        self.commands = defaultdict(list)
+        self.cursor_blink()
+
+    def on_key_press(self, event):
+        if event.keysym == 'Return' and len(self.buffer) > len(self.prefix):
+            self.texts.append(self.last_text)
+            span = self.text_line_span()
+            self.last_text = None
+            self.line_pos = (5, self.line_pos[1] + 15 * span)
+            self.call_command('onreturn', self.buffer[len(self.prefix):])
+            self.buffer = self.prefix
+            self.text_update()
+            return
+
+        if event.keysym == 'Tab':
+            self.call_command('ontab', self)
+            return
+
+        if event.keysym == 'BackSpace' and len(self.buffer) > len(self.prefix):
+            self.buffer = self.buffer[:-1]
+            self.text_update()
+            return
+
+        if event.char.strip() or event.keysym == 'space':
+            self.buffer += event.char
+            self.text_update()
+            return
+
+    def on_configure(self, event):
+        padding = 4
```

Comment:
On Ubuntu 20.04, GNOME, Python 3.8.2, a padding value of 4 makes the widget disappear until manually resized. A padding size of 2 solves this issue. I understand that this is here for stretching the widget to the size of its master, but, speaking from personal experience, this introduces head-aches like this when working with Canvases.
This code should be tested on different platforms in order to make sure that the widget does not resize itself out of visibility.

Reply:
On windows, a padding of 2 makes the widget grow indefinitely